tests/lib/rules/display-name.js

```diff
@@ -672,6 +688,94 @@ ruleTester.run('display-name', rule, {
     errors: [{
       message: 'Component definition is missing display name'
     }]
+  }, {
+    code: `
+      import React from 'react'
+
+      export const ComponentWithMemo = React.memo(({ world }) => {
+        return <div>Hello {world}</div>
+      })
+    `,
+    parser: parsers.BABEL_ESLINT,
+    errors: [{
+      message: 'Component definition is missing display name'
+    }]
+  }, {
+    code: `
+      import React from 'react'
+
+      export const ComponentWithMemo = React.memo(function() {
+        return <div>Hello {world}</div>
+      })
+    `,
+    parser: parsers.BABEL_ESLINT,
+    errors: [{
+      message: 'Component definition is missing display name'
+    }]
+  }, {
+    code: `
+      import React from 'react'
+
+      export const ComponentWithForwardRef = React.forwardRef(({ world }) => {
+        return <div>Hello {world}</div>
+      })
+    `,
+    parser: parsers.BABEL_ESLINT,
+    errors: [{
+      message: 'Component definition is missing display name'
+    }]
+  }, {
+    code: `
+      import React from 'react'
+
+      export const ComponentWithForwardRef = React.forwardRef(function({ world }) {
+        return <div>Hello {world}</div>
+      })
+    `,
+    parser: parsers.BABEL_ESLINT,
+    errors: [{
+      message: 'Component definition is missing display name'
+    }]
+  }, {
+    // This test shows buggy behaviour but the outcome for the user is kinda
+    // right if you squint at it
+    // In this test the plugin identifies two components instead of one.
+    // React is unable to work out the displayName for this code so we're
+    // reporting the right thing, just doing it twice.
+    // I think ideally this should only report one error.
+    code: `
+      import React from 'react'
+
+      export const ComponentWithMemoAndForwardRef = React.memo(
+        React.forwardRef(({ world }) => {
```

Comment:
the thing inside `forwardRef` is not a component, and should not be identified as such. `forwardRef` callbacks take props and `ref`, which isn't what components take.

Reply:
Gotcha, thanks for the clarification :)

In that case I think there is something off with the Component detection, as this testcase identifies two components (as found by doing a `console.log(components.list())` in the `Program:exit` function in the display-name rule). It sounds like it should only identify one - possibly the outer call to `React.memo`. This also occurs on master too.

I don't understand how the internals of component detection work to be confident in modifying it so only a single Component is discovered. It looks like 98d4bf31 did some work towards this.